capsules/extra/src/nmea.rs

```diff
@@ -0,0 +1,126 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
+//! Provides userspace with access to NMEA sentances.
+
+use core::cell::Cell;
+use kernel::errorcode::into_statuscode;
+use kernel::grant::{AllowRoCount, AllowRwCount, Grant, UpcallCount};
+use kernel::hil;
+use kernel::hil::i2c;
+use kernel::processbuffer::WriteableProcessBuffer;
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::utilities::cells::TakeCell;
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Nmea as usize;
+
+#[derive(Clone, Copy, PartialEq, Default)]
+enum Operation {
+    #[default]
+    None,
+    Read,
+}
+
+#[derive(Default)]
+pub struct App {
+    operation: Operation,
+}
+
+pub struct Nmea<'a> {
+    driver: &'a dyn hil::sensors::NmeaDriver<'a>,
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<1>>,
+    busy: Cell<bool>,
+    buffer: TakeCell<'static, [u8]>,
+}
+
+impl<'a> Nmea<'a> {
+    pub fn new(
+        driver: &'a dyn hil::sensors::NmeaDriver<'a>,
+        grant: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<1>>,
+        buffer: &'static mut [u8],
+    ) -> Nmea<'a> {
+        Nmea {
+            driver,
+            apps: grant,
+            busy: Cell::new(false),
+            buffer: TakeCell::new(buffer),
+        }
+    }
+}
+
+impl hil::sensors::NmeaClient for Nmea<'_> {
+    fn callback(&self, buffer: &'static mut [u8], len: usize, status: Result<(), i2c::Error>) {
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, kernel_data| {
+                if app.operation == Operation::Read {
+                    if status.is_err() {
+                        kernel_data
+                            .schedule_upcall(0, (into_statuscode(Err(ErrorCode::FAIL)), 0, 0))
+                            .ok();
+                    }
+
+                    let _ = kernel_data.get_readwrite_processbuffer(0).and_then(|dest| {
+                        dest.mut_enter(|dest| {
+                            let copy_len = dest.len().max(len);
```

Comment:
Shouldn't this be the _minimum_ of `buffer.len()`, `dest.len()`, and `len`? Otherwise, the next line could panic during the copy.